pkg/bid/bid.go

```diff
@@ -0,0 +1,113 @@
+package bid
+
+import (
+	"fmt"
+	"strings"
+
+	v2 "github.com/conductorone/baton-sdk/pb/c1/connector/v2"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	BidPrefix            = "bid"
+	ResourceBidPrefix    = "r"
+	EntitlementBidPrefix = "e"
+	GrantBidPrefix       = "g"
+)
+
+type BID interface {
+	proto.Message
+	// TODO: be smarter about this interface.
+	GetAnnotations() []*anypb.Any
+}
+
+func MakeBid(b BID) (string, error) {
+	switch bType := b.(type) {
+	case *v2.Resource:
+		return makeResourceBid(bType)
+	case *v2.Entitlement:
+		return makeEntitlementBid(bType)
+	case *v2.Grant:
+		return makeGrantBid(bType)
+	}
+	return "", fmt.Errorf("unknown bid type: %T", b)
+}
+
+func MustMakeBid(b BID) string {
+	str, err := MakeBid(b)
+	if err != nil {
+		panic(err)
+	}
+	return str
+}
+
+// escape colon, slash, and backslash.
+func escapeParts(str string) string {
+	escaped := strings.ReplaceAll(str, "\\", "\\\\")
+	escaped = strings.ReplaceAll(escaped, "/", "\\/")
+	escaped = strings.ReplaceAll(escaped, ":", "\\:")
+	return escaped
+}
+
+func resourcePartToStr(r *v2.Resource) (string, error) {
+	rid := r.GetId()
+	resourceType := escapeParts(rid.GetResourceType())
+	resource := escapeParts(rid.GetResource())
+	if resourceType == "" || resource == "" {
+		return "", fmt.Errorf("resource type or id is empty")
+	}
+	if r.ParentResourceId == nil {
+		return fmt.Sprintf("%s/%s", resourceType, resource), nil
```

Comment:
minor: string.Join as a pattern vs sprintf as a pattern -- most of these are something a stirng.Join can do